File: src/autorestart.rs
use crate::runtime_data::s5ciRuntimeData;
use crate::s5ci_config::s5ciConfig;
use s5ci::now_naive_date_time;
use std::time::SystemTime;

pub struct AutorestartState {
    pub config_mtime: Option<SystemTime>,
    pub exe_mtime: Option<SystemTime>,
}

fn restart_ourselves() {
    use std::env;
    use std::process;
    let argv_real: Vec<String> = env::args().collect();
    let err = exec::Command::new(&argv_real[0])
        .args(&argv_real[1..])
        .exec();
    // normally not reached
    println!("Error: {}", err);
    process::exit(1);
}

fn get_mtime(fname: &str) -> Option<std::time::SystemTime> {
    let mtime = std::fs::metadata(fname).ok().map(|x| x.modified().unwrap());
    mtime
}

fn file_changed_since(fname: &str, since: Option<std::time::SystemTime>) -> bool {
    use std::time::{Duration, SystemTime};
    let new_mtime = get_mtime(fname);
    let few_seconds = Duration::from_secs(10);
    if let (Some(old_t), Some(new_t)) = (since, new_mtime) {
        if new_t.duration_since(old_t).unwrap_or(few_seconds) > few_seconds {
            return true;
        }
    }
    // be conservative if we didn't have either of mtimes
    return false;
}

pub fn autorestart_init(config: &s5ciConfig, rtdt: &s5ciRuntimeData) -> AutorestartState {
    let config_mtime = get_mtime(&rtdt.config_path);
    let exe_mtime = get_mtime(&rtdt.real_s5ci_exe);
    AutorestartState {
        config_mtime,
        exe_mtime,
    }
}

pub fn autorestart_check(config: &s5ciConfig, rtdt: &s5ciRuntimeData, ars: &AutorestartState) {
    if config.autorestart.on_config_change
        && file_changed_since(&rtdt.config_path, ars.config_mtime)
    {
        println!(
            "Config changed, attempt restart at {}...",
            now_naive_date_time()
        );
        restart_ourselves();
    }
    if config.autorestart.on_exe_change && file_changed_since(&rtdt.real_s5ci_exe, ars.exe_mtime) {
        println!(
            "Executable changed, attempt restart at {}... ",
            now_naive_date_time()
        );
        restart_ourselves();
    }
}

File: go-s5ci/autorestart.go
package main

import (
	"fmt"
	"log"
	"os"
	"syscall"
)

func GetFileMtimestamp(filename string) int {
	file, err := os.Stat(filename)

	if err != nil {
		fmt.Println("error while doing stat on file ", filename, ":", err)
		return 0
	}

	modifiedtime := int(file.ModTime().Unix())
	return modifiedtime
}

func FileChangedSince(filename string, since int) bool {
	mtime := GetFileMtimestamp(filename)
	delta := 10
	if mtime > 0 && since > 0 {
		if mtime-since > delta {
			return true
		}
	}
	return false
}

type AutorestartState struct {
	ConfigMtime int
	ExeMtime    int
}

func AutorestartInit(c *S5ciConfig, rtdt *S5ciRuntimeData) AutorestartState {
	// log.Printf("exe: %s config: %s", rtdt.RealS5ciExe, rtdt.ConfigPath)
	config_mtime := GetFileMtimestamp(rtdt.ConfigPath)
	exe_mtime := GetFileMtimestamp(rtdt.RealS5ciExe)
	return AutorestartState{ConfigMtime: config_mtime, ExeMtime: exe_mtime}
}

func AutorestartCheck(c *S5ciConfig, rtdt *S5ciRuntimeData, ars *AutorestartState) {
	if c.Autorestart.On_Config_Change &&
		FileChangedSince(rtdt.ConfigPath, ars.ConfigMtime) {
		log.Printf("Config changed, attempt restart")
		S5ciEvent("autorestart-config")
		RestartOurselves(c, rtdt)
	}

	if c.Autorestart.On_Exe_Change && FileChangedSince(rtdt.RealS5ciExe, ars.ExeMtime) {
		log.Printf("Executable changed, attempt restart")
		S5ciEvent("autorestart-exe")
		RestartOurselves(c, rtdt)
	}
}

func RestartOurselves(c *S5ciConfig, rtdt *S5ciRuntimeData) {
	argv0 := rtdt.RealS5ciExe
	argv := append([]string{}, os.Args...)
	new_env := append([]string{}, os.Environ()...)
	syscall.Exec(argv0, argv, new_env)
}
